odl/operator/fn_ops.py

```diff
@@ -0,0 +1,406 @@
+        if self.indices.ndim > 1:
+            # Strides = increment in linear indices per axis
+            strides = np.concatenate((np.cumprod(domain.shape[1:])[::-1], [1]))
+            self.__flat_indices = np.sum(self.indices * strides[:, None], axis=0)
+        
+        range = fn(self.flat_indices.size, dtype=domain.dtype)
+
+        super().__init__(domain, range, linear=True)
+
+    @property
+    def indices(self):
+        """indices where to sample the function."""
+        return self.__indices
+
+    @property
+    def flat_indices(self):
+        """indices where to sample the function."""
+        return self.__flat_indices
+
+    def _call(self, x, out=None):
+        """Collect indices weighted with the cell volume."""
+        c = getattr(self.domain, 'cell_volume', 1.0) # TODO: This only works for a subset of function spaces
+        if out is None:
+            out = x[self.flat_indices]*c
+        else:
+            out[:] = x[self.flat_indices]*c
+        return out
+
+    @property
+    def adjoint(self):
+        """Adjoint of the sampling is 'zero filling'.
+
+        Returns
+        -------
+        adjoint : `ZerofillingOperator`
+        """
+        return ZerofillingOperator(self.domain, self.indices)
+
+    def __repr__(self):
+        """Return ``repr(self)``."""
+        return '{}({!r}, {!r})'.format(self.__class__.__name__,
+                                       self.domain, self.indices)
+
+    def __str__(self):
+        """Return ``str(self)``."""
+        return '{} * I'.format(self.indices)
+
+class ZerofillingOperator(Operator):
+
+    """Operator of multiplication with a scalar.
+
+        ``ZerofillingOperator(f)(x) == sum_{y in I, y == x} f(y), 0 if x not in I``
+    """
+
+    def __init__(self, range, indices):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        range : `LinearSpace` or `Field`
+            Set of elements into which this operator maps.
+        sample_indices : list
+            List of indices (I in the formula above)
+
+        Examples
+        --------
+        >>> X = odl.uniform_discr(min_pt=[0, 0], max_pt=[1, 1], shape=[2, 2])
+        >>> x = X.one()
+        >>> ind = list()
+        >>> ind.append(range(2))
+        >>> ind.append([1,1])
+        >>> A = odl.operator.SamplingOperator(X, ind)
+        >>> A.adjoint(A(x))
+        uniform_discr([0.0, 0.0], [1.0, 1.0], (2, 2)).element([[0.0, 0.25], [0.0, 0.25]])
+        >>> A.adjoint(A(x))  # Out-of-place
+        uniform_discr([0.0, 0.0], [1.0, 1.0], (2, 2)).element([[0.0, 0.25], [0.0, 0.25]])
+
+        >>> X = odl.uniform_discr(min_pt=[-1, -1], max_pt=[1, 1], shape=[4, 3])
+        >>> x = X.element(range(1,X.size+1))
+        >>> ind = list()
+        >>> ind.append(range(3))
+        >>> ind.append(range(3))
+        >>> A = odl.operator.SamplingOperator(X, ind)
+        >>> A.adjoint(A(x)).inner(x) - A(x).inner(A(x))
+        -1.7763568394002505e-15
```

Comment:
Approximate values like this cause bugs and failing tests, I'd prefer doing an approximate test, e.g.

```
>>> abs(A.adjoint(A(x)).inner(x) - A(x).inner(A(x))) < 1e-10
True
```